### PR TITLE
Create link services #LMR-558

### PR DESCRIPTION
**Pull request by kubedo8**

### CODE REVIEW

File: api/src/main/java/io/lumeer/api/dto/JsonQuery.java [hunk 1]
```diff
@@ -111,31 +158,43 @@ public boolean equals(final Object o) {
       if (this == o) {
          return true;
       }
-      if (!(o instanceof Query)) {
+      if (!(o instanceof JsonQuery)) {
```
Comment (livthomas) [inline api/src/main/java/io/lumeer/api/dto/JsonQuery.java]: `Query` interface was deliberately used here in order to be able to compare the instance with `MorphiaQuery` as well. It may not be needed anymore but I still wonder why this change?

File: api/src/main/java/io/lumeer/api/model/LinkInstance.java [hunk 2]
```diff
@@ -19,25 +19,94 @@
...
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 
-public interface LinkInstance {
+public class LinkInstance {
+
+   public static final String ID = "_ID";
```
Comment (livthomas) [inline api/src/main/java/io/lumeer/api/model/LinkInstance.java]: Why do we define `_id` field here? This class should not be mapped to MongoDB in any way.

File: api/src/main/java/io/lumeer/api/model/LinkInstance.java [hunk 3]
```diff
@@ -19,25 +19,94 @@
...
+   private String id;
+   private String linkTypeId;
+   private List<String> documentIds;
+   private Map<String, Object> data;
+
+   @JsonCreator
+   public LinkInstance(@JsonProperty(LINK_TYPE_ID) final String linkTypeId,
+         @JsonProperty(DOCUMENTS_IDS) final List<String> documentIds,
+         @JsonProperty(DATA) final Map<String, Object> data) {
+      this.id = id;
```
Comment (livthomas) [inline api/src/main/java/io/lumeer/api/model/LinkInstance.java]: Where does this `id` come from?

File: api/src/main/java/io/lumeer/api/model/LinkType.java [hunk 4]
```diff
@@ -19,24 +19,94 @@
...
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.util.Collections;
 import java.util.List;
 
-public interface LinkType {
+public class LinkType {
+
+   public static final String ID = "_id";
```
Comment (livthomas) [inline api/src/main/java/io/lumeer/api/model/LinkType.java]: The same thing here. If you want to refer to this as JSON field mapping, it should be `id` instead.

File: api/src/main/java/io/lumeer/api/model/LinkType.java [hunk 5]
```diff
@@ -19,24 +19,94 @@
...
+
+   private String id;
+   private String name;
+   private List<String> collectionIds;
+   private List<JsonAttribute> attributes;
+
+   @JsonCreator
+   public LinkType(@JsonProperty(NAME) final String name,
+         @JsonProperty(COLLECTION_IDS) final List<String> collectionIds,
+         @JsonProperty(ATTRIBUTES) final List<JsonAttribute> attributes) {
```
Comment (livthomas) [inline api/src/main/java/io/lumeer/api/model/LinkType.java]: We would probably want to map `id` field coming from the front-end as well, or not?

File: lumeer-core/src/main/java/io/lumeer/core/facade/LinkInstanceFacade.java [hunk 6]
```diff
@@ -19,27 +19,71 @@
...
+import io.lumeer.storage.api.dao.LinkInstanceDao;
+import io.lumeer.storage.api.dao.LinkTypeDao;
 
+import java.util.ArrayList;
+import java.util.HashSet;
 import java.util.List;
+import java.util.Set;
+import javax.inject.Inject;
 
 public class LinkInstanceFacade {
```
Comment (livthomas) [inline lumeer-core/src/main/java/io/lumeer/core/facade/LinkInstanceFacade.java]: Scope is not defined here and so `@Dependent` is used by default. We should rather explicitly define scope in order to prevent confusion.

File: lumeer-core/src/main/java/io/lumeer/core/facade/LinkInstanceFacade.java [hunk 7]
```diff
@@ -19,27 +19,71 @@
...
+   private LinkTypeDao linkTypeDao;
+
+   @Inject
+   private CollectionDao collectionDao;
+
+   @Inject
+   private LinkInstanceDao linkInstanceDao;
+
+   @Inject
+   protected PermissionsChecker permissionsChecker;
```
Comment (livthomas) [inline lumeer-core/src/main/java/io/lumeer/core/facade/LinkInstanceFacade.java]: You don't need to inject such common things. Just extend `AbstractFacade`.

File: lumeer-core/src/main/java/io/lumeer/core/facade/LinkInstanceFacade.java [hunk 8]
```diff
@@ -19,27 +19,71 @@
...
+      linkInstanceDao.deleteLinkInstance(id);
+   }
+
+   public List<LinkInstance> getLinkInstances(Query query) {
+      return linkInstanceDao.getLinkInstances(query);
    }
 
-   public List<LinkInstance> getLinkInstances(SearchQuery query) {
-      throw new UnsupportedOperationException();
+   private void checkLinkInstancePermission(Set<String> collectionIds) {
```
Comment (livthomas) [inline lumeer-core/src/main/java/io/lumeer/core/facade/LinkInstanceFacade.java]: Why not using `Collection<String>` here so we don't have to convert it with almost every use?

File: lumeer-core/src/main/java/io/lumeer/core/facade/LinkTypeFacade.java [hunk 9]
```diff
@@ -19,27 +19,62 @@
...
 public class LinkTypeFacade {
 
+   @Inject
+   private LinkTypeDao linkTypeDao;
+
+   @Inject
+   private CollectionDao collectionDao;
+
+   @Inject
+   protected PermissionsChecker permissionsChecker;
```
Comment (livthomas) [inline lumeer-core/src/main/java/io/lumeer/core/facade/LinkTypeFacade.java]: The same applies here as well. We can just extend `AbstractFacade` and get rid of injecting common stuff.

File: lumeer-core/src/main/java/io/lumeer/core/facade/LinkTypeFacade.java [hunk 10]
```diff
@@ -19,27 +19,62 @@
...
+      linkTypeDao.deleteLinkType(id);
+   }
+
+   public List<LinkType> getLinkTypes(Query query) {
+      return linkTypeDao.getLinkTypes(query);
    }
 
-   public List<LinkType> getLinkTypes(SearchQuery query) {
-      throw new UnsupportedOperationException();
+   private void checkLinkTypePermission(Set<String> collectionIds) {
```
Comment (livthomas) [inline lumeer-core/src/main/java/io/lumeer/core/facade/LinkTypeFacade.java]: The same applies here. `Collection<String>` would save some unnecessary conversions.

File: lumeer-remote/src/main/java/io/lumeer/remote/rest/LinkInstanceService.java [hunk 11]
```diff
@@ -19,29 +19,67 @@
...
    public LinkInstance createLinkInstance(LinkInstance linkInstance) {
-      throw new UnsupportedOperationException();
+      return linkInstanceFacade.createLinkInstance(linkInstance);
    }
 
-   public LinkInstance updateLinkInstance(String id, LinkInstance linkInstance) {
-      throw new UnsupportedOperationException();
+   @PUT
+   @Path("{linkIstanceId}")
+   public LinkInstance updateLinkInstance(@PathParam("linkIstanceId") String id, LinkInstance linkInstance) {
```
Comment (livthomas) [inline lumeer-remote/src/main/java/io/lumeer/remote/rest/LinkInstanceService.java]: Two typos detected! :D

File: lumeer-remote/src/main/java/io/lumeer/remote/rest/LinkInstanceService.java [hunk 12]
```diff
@@ -19,29 +19,67 @@
...
+      return linkInstanceFacade.updateLinkInstance(id, linkInstance);
    }
 
-   public void deleteLinkInstance(String id) {
-      throw new UnsupportedOperationException();
+   @DELETE
+   @Path("{linkIstanceId}")
+   public Response deleteLinkInstance(@PathParam("linkIstanceId") String id) {
```
Comment (livthomas) [inline lumeer-remote/src/main/java/io/lumeer/remote/rest/LinkInstanceService.java]: And two more...

File: lumeer-storage/lumeer-storage-api/src/main/java/io/lumeer/storage/api/dao/LinkInstanceDao.java [hunk 13]
```diff
@@ -20,20 +20,29 @@
...
+   void deleteLinkInstance(String id);
 
-   void deleteLinkInstance(String linkInstanceId);
+   void deleteLinkInstances(Query query);
 
-   void deleteLinkInstances(SearchQuery query);
+   LinkInstance getLinkInstance(String id);
 
-   List<LinkInstance> getLinkInstances(SearchQuery query);
+   List<LinkInstance> getLinkInstances(Query query);
```
Comment (livthomas) [inline lumeer-storage/lumeer-storage-api/src/main/java/io/lumeer/storage/api/dao/LinkInstanceDao.java]: I think we use `SearchQuery` on DAO level.

File: lumeer-storage/lumeer-storage-api/src/main/java/io/lumeer/storage/api/dao/CollectionDao.java [hunk 14]
```diff
@@ -42,6 +42,8 @@
 
    Collection getCollectionByCode(String code);
 
+   List<Collection> getCollectionByIds(List<String> ids);
```
Comment (livthomas) [inline lumeer-storage/lumeer-storage-api/src/main/java/io/lumeer/storage/api/dao/CollectionDao.java]: `getCollectionsByIds`

File: lumeer-storage/lumeer-storage-mongodb/src/main/java/io/lumeer/storage/mongodb/dao/project/MorphiaCollectionDao.java [hunk 15]
```diff
@@ -92,6 +92,13 @@ public Collection getCollectionByCode(final String code) {
       return collection;
    }
 
+   @Override
+   public List<Collection> getCollectionByIds(final List<String> ids) {
```
Comment (livthomas) [inline lumeer-storage/lumeer-storage-mongodb/src/main/java/io/lumeer/storage/mongodb/dao/project/MorphiaCollectionDao.java]: What about using `Collection<String>` here? I can see that we sometimes need to pass `Set` here and it doesn't really matter what collection type it is inside this method.

File: api/src/main/java/io/lumeer/api/dto/JsonQuery.java (hunk 1 above)
Comment (kubedo8) [reply api/src/main/java/io/lumeer/api/dto/JsonQuery.java]: This was auto-generated. I will fix that.

File: api/src/main/java/io/lumeer/api/model/LinkInstance.java (hunk 2 above)
Comment (kubedo8) [reply api/src/main/java/io/lumeer/api/model/LinkInstance.java]: It is unused. I will remove it.

File: api/src/main/java/io/lumeer/api/model/LinkType.java (hunk 5 above)
Comment (kubedo8) [reply api/src/main/java/io/lumeer/api/model/LinkType.java]: Yes of course

File: lumeer-core/src/main/java/io/lumeer/core/facade/LinkInstanceFacade.java (hunk 7 above)
Comment (kubedo8) [reply lumeer-core/src/main/java/io/lumeer/core/facade/LinkInstanceFacade.java]: AbstractFacade contains lot of unnecessary things, but Ok, I will extend it.

File: lumeer-core/src/main/java/io/lumeer/core/facade/LinkTypeFacade.java (hunk 10 above)
Comment (kubedo8) [reply lumeer-core/src/main/java/io/lumeer/core/facade/LinkTypeFacade.java]: But I need distinct values here

File: lumeer-storage/lumeer-storage-mongodb/src/main/java/io/lumeer/storage/mongodb/codecs/LinkTypeCodec.java [hunk 16]
```diff
@@ -40,6 +40,11 @@
 
 public class LinkTypeCodec implements CollectibleCodec<LinkType> {
 
+   public static final String ID = "_id";
+   public static final String NAME = "name";
+   public static final String COLLECTION_IDS = "collectionIds";
+   public static final String ATTRIBUTES = "attributes";
```
Comment (livthomas) [inline lumeer-storage/lumeer-storage-mongodb/src/main/java/io/lumeer/storage/mongodb/codecs/LinkTypeCodec.java]: +1 on defining it here

File: lumeer-storage/lumeer-storage-mongodb/src/main/java/io/lumeer/storage/mongodb/dao/project/MongoLinkInstanceDao.java [hunk 17]
```diff
@@ -35,7 +36,7 @@
    public void createLinkInstanceRepository(Project project) {
       database.createCollection(databaseCollectionName(project));
 
-      MongoCollection<LinkInstance> projectCollection = databaseCollection();
+      MongoCollection<Document> projectCollection = database.getCollection(databaseCollectionName(project));
```
Comment (livthomas) [inline lumeer-storage/lumeer-storage-mongodb/src/main/java/io/lumeer/storage/mongodb/dao/project/MongoLinkInstanceDao.java]: Why this change?

Comment (kubedo8) [reply lumeer-storage/lumeer-storage-mongodb/src/main/java/io/lumeer/storage/mongodb/dao/project/MongoLinkInstanceDao.java]: It's same as in other Dao-s. It's cause test exception, because when we are creating repository on Project creation, we don't have set `Project` property here (but it's correct behavior).
